### PR TITLE
LSP breakpoint placement

### PR DESCRIPTION
**Pull request by tmat**
Implements https://github.com/dotnet/roslyn/issues/59455
LSP: https://github.com/microsoft/language-server-protocol/pull/1401

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        public override bool MutatesSolutionState => false;
+        public override bool RequiresLSPSolution => true;
+
+        public override LSP.TextDocumentIdentifier? GetTextDocumentIdentifier(LSP.ValidateBreakableRangeParams request)
+            => request.TextDocument;
+
+        public override async Task<LSP.Range?> HandleRequestAsync(LSP.ValidateBreakableRangeParams request, RequestContext context, CancellationToken cancellationToken)
+        {
+            var document = context.Document;
+            if (document == null)
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs:56]: Potentially should throw here if we weren't able to get a document for the request

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            Range = range;
+        }
+    }
+}
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    [ExportRoslynLanguagesLspRequestHandlerProvider, Shared]
+    [ProvidesMethod("textDocument/validateBreakableRange")]
+    internal sealed class ValidateBreakableRangeHandler : AbstractStatelessRequestHandler<LSP.ValidateBreakableRangeParams, LSP.Range?>
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs:37]: would be good to add a test that at least validates that we can call this and get a response.  Probably not much more needed here assuming we already have test cases for the IBreakpointResolutionService

**Review by dibarbet — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs (hunk 1 above)
Comment (tmat) [reply src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs]: Done. Why is context.Document nullable if we have to have it?

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs (hunk 2 above)
Comment (tmat) [reply src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs]: Added tests

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs (hunk 1 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs]: Not all requests provide a document, for example workspace/symbol or workspace diagnostics

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.CodeAnalysis.Host.Mef;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    [ExportRoslynLanguagesLspRequestHandlerProvider, Shared]
+    [ProvidesMethod(LSP.VSInternalMethods.TextDocumentValidateBreakableRangeName)]
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs:18]: This was refactored in between you creating this PR and now 😄 
Now the attributes need to look like this - https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Protocol/Handler/FoldingRanges/FoldingRangesHandler.cs#L20

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    internal sealed class ValidateBreakableRangeHandler : AbstractStatelessRequestHandler<LSP.VSInternalValidateBreakableRangeParams, LSP.Range?>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public ValidateBreakableRangeHandler()
+        {
+        }
+
+        public override string Method => LSP.VSInternalMethods.TextDocumentValidateBreakableRangeName;
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs:27]: this property on the base has been removed and isn't needed

**Review by dibarbet — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs (hunk 1 above)
Comment (davidwengier) [reply src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs]: To expand on this, `RequiresLSPSolution` being `true` and  and `GetTextDocumentIdentifier` not returning null is a contract that says `context.Document` can't be null. It's just annoying to express that in the API :)

**Review by davidwengier — APPROVED** [review]
You can tell who designed the LSP spec when the handler implementation is this simple 😛

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            => request.TextDocument;
+
+        public override async Task<LSP.Range?> HandleRequestAsync(LSP.VSInternalValidateBreakableRangeParams request, RequestContext context, CancellationToken cancellationToken)
+        {
+            var document = context.Document;
+            Contract.ThrowIfNull(document);
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var span = ProtocolConversions.RangeToTextSpan(request.Range, text);
+            var breakpointService = document.Project.LanguageServices.GetRequiredService<IBreakpointResolutionService>();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs:40]: will this break for non-c#/vb @dibarbet shoudl we make this GetService and then return null if it's not available?

**Comment by genlu** [comment]
This change has broken ngen in[ our isnertion](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS/pullrequest/382228)

```
02/25/2022 10:03:22.907 [4344]: Failed to load dependency Microsoft.VisualStudio.LanguageServer.Protocol.Internal of assembly Microsoft.CodeAnalysis.EditorFeatures, Version=4.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 because of the following error : The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
02/25/2022 10:03:22.907 [4344]: Failed to load dependency Microsoft.VisualStudio.LanguageServer.Protocol of assembly Microsoft.CodeAnalysis.EditorFeatures, Version=4.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 because of the following error : The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
02/25/2022 10:03:22.923 [4344]: Failed to load dependency Microsoft.VisualStudio.LanguageServer.Protocol.Extensions of assembly Microsoft.CodeAnalysis.EditorFeatures, Version=4.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 because of the following error : The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

```

File: src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs (hunk 5 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Breakpoints/ValidateBreakableRangeHandler.cs:40]: I don't think its necessary - we should just not advertise the capability for other scenarios

**Comment by dibarbet** [comment]
Looks like 17.2.3 insertion is not merged yet - https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/382251

**Pull request merged**